### PR TITLE
sql: tweak some char casting comments + code

### PR DESCRIPTION
**Pull request by benesch**
See commits for details.

### Motivation

   * This PR refactors existing code.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

**Comment by benesch** [comment]
Ahh, ok, I found the reason for the coercion special case, and I think a tweak to a _different_ special case makes the former evaporate.

**Comment by benesch** [comment]
Hrm, the failures here suggest that this PR actually needs to wait for #10911 to merge. Marking as a draft in the meantime.

**Comment by benesch** [comment]
Ok, I think I finally sorted this out. Let's see what CI thinks.

**Comment by benesch** [comment]
@sploiselle this is ready for a look!

File: test/sqllogictest/type-promotion.slt [hunk 1]
```diff
@@ -280,7 +280,7 @@ EXPLAIN RAW PLAN FOR
 ----
 %0 =
 | Constant ()
-| Map lower(chartostr(strtochar("a")))
+| Map lower(chartostr(strtochar(strtochar("a"))))
```
Comment (sploiselle) [inline test/sqllogictest/type-promotion.slt:283]: The doubled call to `strtochar` are a little mind-bending to me--do you have immediate intuition why the intermediate call's getting injected where it wasn't before? Glad to look at this if the answer doesn't come to you in like 10 seconds.

**Comment by benesch** [comment]
Ah yeah, it’s the result of this change: https://github.com/MaterializeInc/materialize/pull/10909/files#diff-b945a38f3da415078dc2535b510ccea6e3d75cd0b441e34426418a3543cb702bR793

String literal coercion now happens in two steps. First the literal is explicitly cast to the unmodified type, then it’s implicitly cast to the modified type. AFAICT this is exactly what Postgres does, so the duplicate call is expected!

**Review by sploiselle — APPROVED** [review]
Ah, it was stripping off the typmod and then reapplying it that threw me off. tysm for this incredibly subtle fix.

**Comment by benesch** [comment]
Yeah it’s a little wild, isn’t it? It all makes sense in retrospect… implicit casts error, because you don’t want to silently truncate data… explicit casts can truncate data because they are explicit. But then you’re in a pickle with string literal coercion; you want to do casts that wouldn’t normally apply, because you have an untyped string literal rather than a fixed type expression, but you still don’t want to silently truncate data, so you have to do the conversion to the base type explicitly but apply the typemods implicitly.

**Pull request merged**